File: mrn_ws/src/upc_mrn/src/odom_new.cpp
#include "ros/ros.h"
#include "nav_msgs/Odometry.h"
#include "sensor_msgs/JointState.h"
#include <tf/transform_datatypes.h>

class OdomNew
{
  public:
    OdomNew()
    {
      this->joint_state_sub  = n.subscribe("joint_states", 1, &OdomNew::joint_state_callback, this);
      this->received_joint_state = false;
      
      this->odom_sub  = n.subscribe("odom", 1, &OdomNew::odom_callback, this);
      this->received_odom = false;
      this->get_x_y_th    = true;
      
      this->odom_pub = n.advertise<nav_msgs::Odometry>("odom_new", 1);
      this->x=0.0;
      this->y=0.0;
      this->th=0.0;
      
      this->radius=0.035;
      this->wheel_distance=0.23;
      this->last_time = ros::Time::now();
    }

    void joint_state_callback(const sensor_msgs::JointState::ConstPtr& msg)
    {
      this->joint_state_msg = *msg;
      this->received_joint_state=true;
    }
    
    void odom_callback(const nav_msgs::Odometry::ConstPtr& msg)
    {
      if(!this->received_odom)
      {
        ROS_INFO("Received input odom");
        this->input_odom_msg = *msg;
        this->received_odom=true;
      }
    }

    void compute_and_publish_odom()
    {
      //ROS_INFO("compute_and_publish_odom");
      double left_pos, right_pos, left_vel, right_vel;
      if(this->joint_state_msg.position.size()>=2 && this->joint_state_msg.velocity.size()>=2)
      {
        left_pos  = this->joint_state_msg.position[0]; //CAUTION name field should be checked, for now, assuming 0=left, 1=right
        right_pos = this->joint_state_msg.position[1];
        left_vel  = this->joint_state_msg.velocity[0];
        right_vel = this->joint_state_msg.velocity[1];
      }
      else
        ROS_ERROR("Wrong input joint_state size");

      ros::Time time = this->joint_state_msg.header.stamp;
      //ROS_INFO("Received left_pos,right_pos: %f,%f, left_vel,right_vel: %f,%f", left_pos, right_pos, left_vel, right_vel);
      double dt = (time - this->last_time).toSec();
      //ROS_INFO("dt: %fs",dt);
      double vx=0.0;
      double vth=0.0;
      
      //////// TODO //////
      //vx  = 
      //vth = 
      //this->x  = 
      //this->y  = 
      //this->th = 
      //////// TODO //////
      
      this->odom_msg.header.stamp = time;
      this->odom_msg.header.frame_id = "odom";
      this->odom_msg.pose.pose.position.x = this->x;
      this->odom_msg.pose.pose.position.y = this->y;
      this->odom_msg.pose.pose.orientation = tf::createQuaternionMsgFromYaw(this->th);
      //this->odom_msg.pose.covariance=
      this->odom_msg.child_frame_id = "base_footprint_new";
      this->odom_msg.twist.twist.linear.x = vx;
      this->odom_msg.twist.twist.angular.z = vth;
      //this->odom_msg.twist.covariance=
      
      this->odom_pub.publish(this->odom_msg);
      this->last_time=time;
    }
    
    void loop()
    {
      //ROS_INFO("loop");
      if(received_odom && get_x_y_th)
      {
        this->x = this->input_odom_msg.pose.pose.position.x;
        this->y = this->input_odom_msg.pose.pose.position.y;
        this->th = tf::getYaw(this->input_odom_msg.pose.pose.orientation);
        this->last_time = this->input_odom_msg.header.stamp;
        ROS_INFO("Initialization: getting from input odom x,y,th=%f,%f,%f", this->x,this->y,this->th);
        this->get_x_y_th=false;
      }
      
      if(this->received_joint_state)
        this->compute_and_publish_odom();
    }

  private:
    ros::NodeHandle n; 
    ros::Subscriber odom_sub;
    nav_msgs::Odometry input_odom_msg;
    bool received_odom;
    bool get_x_y_th;
    ros::Subscriber joint_state_sub;
    sensor_msgs::JointState joint_state_msg;
    bool received_joint_state;
    ros::Publisher  odom_pub;
    nav_msgs::Odometry odom_msg;
    double x,y,th;
    double radius, wheel_distance;
    ros::Time last_time;
};

int main(int argc, char **argv)
{
  ros::init(argc, argv, "odom_new");
  OdomNew my_odom_new;
  ros::Rate loop_rate(10);
  while (ros::ok())
  {
    my_odom_new.loop();
    ros::spinOnce();
    loop_rate.sleep();
  }
  return 0;
}

File: mrn_ws/src/upc_mrn/src/odom_new.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import JointState
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Point, Pose, Quaternion, Twist, Vector3
import tf
from math import cos, sin

class odom_new():

    def __init__(self):
        rospy.init_node('odom', anonymous=True)
        self.jointstates_sub = rospy.Subscriber("joint_states", JointState, self.joint_state_callback)
        self.joint_state_msg = JointState()
        self.received_joint_state = False
        self.odom_sub = rospy.Subscriber("odom", Odometry, self.odom_callback)
        self.input_odom_msg = Odometry()
        self.received_odom = False
        self.get_x_y_th = True
        self.odom_pub = rospy.Publisher('odom_new', Odometry, queue_size=1)
        self.odom_msg = Odometry()
        self.x = 0
        self.y = 0
        self.th= 0
        self.rate = rospy.Rate(10)
        self.radius=0.035
        self.wheel_distance=0.23
        self.last_time = rospy.Time()

    def joint_state_callback(self,msg):
        self.joint_state_msg = msg
        self.received_joint_state = True
        
    def odom_callback(self,msg):
        if not self.received_odom:
          rospy.loginfo("Received input odom")
          self.input_odom_msg = msg
          self.received_odom = True
        
    def compute_and_publish_odom(self):
        #rospy.loginfo("compute_and_publish_odom")
        size=2
        if(len(self.joint_state_msg.position) >= size and len(self.joint_state_msg.velocity)>= size):
          left_pos  = self.joint_state_msg.position[0] #CAUTION name field should be checked, for now, assuming 0=left, 1=right
          right_pos = self.joint_state_msg.position[1]
          left_vel  = self.joint_state_msg.velocity[0]
          right_vel = self.joint_state_msg.velocity[1]
        else:
          rospy.logerr("Wrong input joint_state size %d<%d",len(self.joint_state_msg.position), size)
        time = self.joint_state_msg.header.stamp
        #rospy.loginfo("Received left_pos,right_pos: %f, %f, left_vel,right_vel: %f,%f", left_pos, right_pos, left_vel, right_vel)
        dt = (time - self.last_time).to_sec()
        #rospy.loginfo("dt: %fs",dt);
        vx=0
        vth=0

        #### TO DO ###
        #vx = 
        #vth = 
        #self.x  =
        #self.y  =
        #self.th =
        #### TO DO ###

        self.odom_msg.header.stamp = time
        self.odom_msg.header.frame_id = "odom"
        q = tf.transformations.quaternion_from_euler(0, 0, self.th)
        self.odom_msg.pose.pose = Pose(Point(self.x, self.y, 0.), Quaternion(*q))
        #self.odom_msg.pose.covariance=
        self.odom_msg.child_frame_id = "base_footprint_new"
        self.odom_msg.twist.twist = Twist(Vector3(vx, 0, 0), Vector3(0, 0, vth))
        #self.odom_msg.twist.covariance=
        self.odom_pub.publish(self.odom_msg)
        self.last_time=time
        
    def loop(self):
        #rospy.loginfo("loop")
        if self.received_odom and self.get_x_y_th:
          self.x = self.input_odom_msg.pose.pose.position.x
          self.y = self.input_odom_msg.pose.pose.position.y
          self.th = tf.transformations.euler_from_quaternion([self.input_odom_msg.pose.pose.orientation.x, self.input_odom_msg.pose.pose.orientation.y, self.input_odom_msg.pose.pose.orientation.z, self.input_odom_msg.pose.pose.orientation.w])[2]
          self.last_time = self.input_odom_msg.header.stamp
          rospy.loginfo("Initialization: getting from input odom x,y,th=%f,%f,%f",self.x,self.y,self.th)
          self.get_x_y_th = False
          
        if self.received_joint_state:
            self.compute_and_publish_odom()
      

if __name__ == '__main__':
    try:
        #Testing our function
        n = odom_new()
        rate = rospy.Rate(10) # 10hz
        while not rospy.is_shutdown():
            n.loop()
            rate.sleep()

    except rospy.ROSInterruptException: pass